src/cache/mod.rs

```diff
@@ -0,0 +1,233 @@
+    where
+        T: Eq + Hash,
+        V: Clone,
+    {
+        self.items
+            .read()
+            .await
+            .get(key)
+            .filter(|&item| !item.expired())
+            .map(|k| k.object.clone())
+    }
+
+    ///
+    pub async fn set(&self, key: T, value: V, custom_duration: Option<Duration>) -> Option<V>
+    where
+        T: Eq + Hash,
+    {
+        self.items
+            .write()
+            .await
+            .insert(
+                key,
+                Item::new(value, custom_duration.or(self.item_duration)),
+            )
+            .map(|item| item.object)
+    }
+
+    ///
+    pub async fn remove_expired(&self)
+    where
+        T: Eq + Hash + Clone,
+    {
+        let expired_keys: Vec<T> = self
+            .items
+            .read()
+            .await
+            .iter()
+            .filter(|(_, item)| item.expired())
+            .map(|(key, _)| key.clone())
+            .collect();
+
+        for key in expired_keys {
+            let _ = self.items.write().await.remove(&key);
+        }
+    }
+
+    ///
+    pub async fn remove(&self, key: &T) -> Option<V>
+    where
+        T: Eq + Hash,
+    {
+        self.items.write().await.remove(key).map(|item| item.object)
+    }
+
+    ///
+    pub async fn clear(&self) {
+        self.items.write().await.clear()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::cache::Cache;
+    use std::time::Duration;
+
+    const KEY: i8 = 0;
+    const VALUE: &str = "VALUE";
+
+    #[tokio::test]
+    async fn set_and_get_value_with_default_duration() {
+        let cache = Cache::new(Some(Duration::from_secs(2)));
+        let _ = cache.set(KEY, VALUE, None).await;
+        let value = cache.get(&KEY).await;
+        match value {
+            Some(value) => assert_eq!(value, VALUE),
+            None => panic!("value was not found in cache"),
+        };
+    }
+
+    #[tokio::test]
+    async fn set_and_get_value_without_duration() {
+        let cache = Cache::new(None);
+        let _ = cache.set(KEY, VALUE, None).await;
+        let value = cache.get(&KEY).await;
+        match value {
+            Some(value) => assert_eq!(value, VALUE),
+            None => panic!("value was not found in cache"),
+        };
+    }
+
+    #[tokio::test]
+    async fn set_and_get_value_with_custom_duration() {
+        let cache = Cache::new(Some(Duration::from_secs(0)));
+        let _ = cache.set(KEY, VALUE, Some(Duration::from_secs(2))).await;
+        let value = cache.get(&KEY).await;
+        match value {
+            Some(value) => assert_eq!(value, VALUE),
+            None => panic!("value was not found in cache"),
+        };
+    }
+
+    #[tokio::test]
+    async fn set_do_not_get_expired_value() {
+        let cache = Cache::new(Some(Duration::from_secs(0)));
+        let _ = cache.set(KEY, VALUE, None).await;
+        let value = cache.get(&KEY).await;
+        if value.is_some() {
+            panic!("found expired value in cache")
```

Comment:
Cold we move panics to asserts please? `assert!(value.is_some())` or similar

Reply:
Mm, yes, missed this one. I copy pasted code :+1:

Reply:
There, fixed.